storage/badger.go

```diff
@@ -0,0 +1,186 @@
+package storage
+
+import (
+	"github.com/dgraph-io/badger/v3"
+	"github.com/pkg/errors"
+	"os"
+)
+
+const (
+	// PerformanceMemTableSize is 3072 MB. Increases the maximum
+	// amount of data we can commit in a single transaction.
+	PerformanceMemTableSize = 3072 << 20
+
+	// PerformanceLogValueSize is 256 MB.
+	PerformanceLogValueSize = 256 << 20
+)
+
+type BadgerDatabase struct {
+	db            *badger.DB
+	opts          badger.Options
+	useWriteBatch bool
+}
+
+func NewBadgerDatabase(opts badger.Options, useWriteBatch bool) *BadgerDatabase {
+	return &BadgerDatabase{
+		db:            nil,
+		opts:          opts,
+		useWriteBatch: useWriteBatch,
+	}
+}
+
+func (bdb *BadgerDatabase) Setup() error {
+	db, err := badger.Open(bdb.opts)
+	if err != nil {
+		return errors.Wrapf(err, "Setup:")
+	}
+	bdb.db = db
+	return nil
+}
+
+func (bdb *BadgerDatabase) Update(fn func(Transaction) error) error {
+	var wb *badger.WriteBatch
+	if bdb.useWriteBatch {
+		wb = bdb.db.NewWriteBatch()
+	}
+
+	err := bdb.db.Update(func(txn *badger.Txn) error {
+		T := NewBadgerTransaction(txn, wb)
+		return fn(T)
+	})
+	if err != nil {
+		return errors.Wrapf(err, "Update:")
+	}
+
+	if bdb.useWriteBatch {
+		err = wb.Flush()
+		if err != nil {
+			return errors.Wrapf(err, "Update: Problem flushing write batch")
+		}
+		wb.Cancel()
+	}
+	return nil
+}
+
+func (bdb *BadgerDatabase) View(fn func(Transaction) error) error {
+	return bdb.db.View(func(txn *badger.Txn) error {
+		T := NewBadgerTransaction(txn, nil)
+		return fn(T)
+	})
+}
+
+func (bdb *BadgerDatabase) Close() error {
+	return bdb.db.Close()
+}
+
+func (bdb *BadgerDatabase) Erase() error {
+	return os.RemoveAll(bdb.opts.Dir)
+}
+
+// ==========================
+// BadgerTransaction
+// ==========================
+
+type BadgerTransaction struct {
+	txn *badger.Txn
+	wb  *badger.WriteBatch
+}
+
+func NewBadgerTransaction(txn *badger.Txn, wb *badger.WriteBatch) *BadgerTransaction {
+	return &BadgerTransaction{
+		txn: txn,
+		wb:  wb,
+	}
+}
+
+func (btx *BadgerTransaction) Set(key []byte, value []byte) error {
+	if btx.wb != nil {
+		return btx.wb.Set(key, value)
+	}
+	return btx.txn.Set(key, value)
+}
+
+func (btx *BadgerTransaction) Delete(key []byte) error {
+	if btx.wb != nil {
+		return btx.wb.Delete(key)
+	}
+	return btx.txn.Delete(key)
+}
+
+func (btx *BadgerTransaction) Get(key []byte) ([]byte, error) {
+	var value []byte
+
+	item, err := btx.txn.Get(key)
+	if err != nil {
+		return value, errors.Wrapf(err, "Get:")
+	}
+	return item.ValueCopy(nil)
```

Comment:
```suggestion
	item, err := btx.txn.Get(key)
	if err != nil {
		return nil, errors.Wrapf(err, "Get:")
	}
	return item.ValueCopy(nil)
```
I don't think we need to declare this `value` variable